File: tools/pwig/demos/cpp_callee/testlib_callee_impl.cpp
// ************************************************************************
// This file implements the callee wrapper classes.

#include <string>
#include "stdafx.h"
#include "stdint.h"
#include "testlib_intf.h"
#include "testlib_callee.h"
#include "testlib_callee_impl.h"

// Name: ProjectGroup
// GUID: 7C12BB43-A6AB-4A52-8B1D-EDD5D94B344B
// Description: ProjectGroup Object

// Constructor:
ProjectGroup::ProjectGroup()
{
  m_Events = new AuxIProjectGroupEvents();
}

// Destructor:
ProjectGroup::~ProjectGroup()
{
  delete m_Events;
}

// Methods:
IProject ProjectGroup::AddProject()
{
  Project * Prj = new Project();
  Prj->m_ProjectGroup = this;
  return (IProject)Prj;
}

void ProjectGroup::RunPeriodic()
{
  // nothing
}

void ProjectGroup::Finalize()
{
  // nothing
}

// Properties:

// Setup event handlers:
void ProjectGroup::SetupOnError(IProjectGroupEvents EventSink, TIProjectGroupEventsOnError EventHandler)
{
  m_Events->SetupOnError(EventSink, EventHandler);
}

void ProjectGroup::SetupOnProgress(IProjectGroupEvents EventSink, TIProjectGroupEventsOnProgress EventHandler)
{
  m_Events->SetupOnProgress(EventSink, EventHandler);
}

void ProjectGroup::Error(TErrorCode ErrorCode, std::wstring ErrorText)
{
  m_Events->OnError(ErrorCode, ErrorText);
}


// Name: Project
// GUID: D96EA22B-D750-4C05-9F32-8C5C8E9F846D
// Description: Project Object

// Constructor:
Project::Project()
{
  // nothing
}

// Destructor:
Project::~Project()
{
  // nothing
}

// Methods:
void Project::Connect()
{
  m_ProjectGroup->Error(errError, L"Cannot connect, not implemented!");
}

void Project::Disconnect()
{
  m_ProjectGroup->Error(errError, L"Cannot disconnect, not implemented!");
}

TBool Project::LoadFromFile(std::wstring Path)
{
  m_ProjectGroup->Error(errError, L"Cannot load from file, not implemented!");
  return bFalse;
}

TBool Project::SaveToFile(std::wstring Path)
{
  m_ProjectGroup->Error(errError, L"Cannot save to file, not implemented!");
  return bFalse;
}

// Properties:
int32_t Project::GetConnectionFRC()
{
  return 0; // not implemented
}

std::wstring Project::GetConnectionString()
{
  return L"No connection, not implemented in library!";
}

File: tools/pwig/demos/pascal_callee/testlib_callee_impl.pas
unit testlib_callee_impl;

{$mode delphi}

interface

uses
  Classes, SysUtils, testlib_callee, testlib_intf;

type

  // Forward declarations:
  TProjectGroup = class;
  TProject = class;

  // Name: ProjectGroup
  // GUID: 7C12BB43-A6AB-4A52-8B1D-EDD5D94B344B
  // Description: ProjectGroup Object

  { TProjectGroup }

  TProjectGroup = class(TObject)
  private
    FEvents: TIProjectGroupEvents;
  protected
    procedure Error(ErrorCode: TErrorCode; const ErrorText: string);
  public
    constructor Create;
    destructor Destroy; override;
    // Methods:
    function AddProject: IProject;
    procedure RunPeriodic;
    procedure Finalize;
    // Properties:
    // Setup event handlers:
    procedure SetupOnError(const EventSink: IProjectGroupEvents; const EventHandler: TIProjectGroupEventsOnError);
    procedure SetupOnProgress(const EventSink: IProjectGroupEvents; const EventHandler: TIProjectGroupEventsOnProgress);
  end;

  // Name: Project
  // GUID: D96EA22B-D750-4C05-9F32-8C5C8E9F846D
  // Description: Project Object

  TProject = class(TObject)
  private
    FProjectGroup: TProjectGroup;
  protected
    property ProjectGroup: TProjectGroup read FProjectGroup write FProjectGroup;
  public
    constructor Create;
    destructor Destroy; override;
    // Methods:
    procedure Connect;
    procedure Disconnect;
    function LoadFromFile(const Path: AnsiString): TBool;
    function SaveToFile(const Path: AnsiString): TBool;
    // Properties:
    function GetConnectionFRC: LongInt;
    function GetConnectionString: AnsiString;
  end;

implementation


{ TProjectGroup }

constructor TProjectGroup.Create;
begin
  FEvents := TIProjectGroupEvents.Create;
end;

destructor TProjectGroup.Destroy;
begin
  FEvents.Free;
  inherited Destroy;
end;

function TProjectGroup.AddProject: IProject;
var
  Prj: TProject;
begin
  Prj := TProject.Create;
  Prj.ProjectGroup := Self;
  Result := IProject(Prj);
end;

procedure TProjectGroup.RunPeriodic;
begin
  // nothing
end;

procedure TProjectGroup.Finalize;
begin
  // nothing
end;

procedure TProjectGroup.SetupOnError(const EventSink: IProjectGroupEvents;
  const EventHandler: TIProjectGroupEventsOnError);
begin
  FEvents.SetupOnError(EventSink, EventHandler);
end;

procedure TProjectGroup.SetupOnProgress(const EventSink: IProjectGroupEvents;
  const EventHandler: TIProjectGroupEventsOnProgress);
begin
  FEvents.SetupOnProgress(EventSink, EventHandler);
end;

procedure TProjectGroup.Error(ErrorCode: TErrorCode; const ErrorText: string);
begin
  FEvents.OnError(ErrorCode, ErrorText);
end;

{ TProject }

constructor TProject.Create;
begin

end;

destructor TProject.Destroy;
begin
  inherited Destroy;
end;

procedure TProject.Connect;
begin
  ProjectGroup.Error(errError, 'Cannot connect, not implemented!');
end;

procedure TProject.Disconnect;
begin
  ProjectGroup.Error(errError, 'Cannot disconnect, not implemented!');
end;

function TProject.LoadFromFile(const Path: AnsiString): TBool;
begin
  ProjectGroup.Error(errError, 'Cannot load from file, not implemented!');
  Result := bFalse;
end;

function TProject.SaveToFile(const Path: AnsiString): TBool;
begin
  ProjectGroup.Error(errError, 'Cannot save to file, not implemented!');
  Result := bFalse;
end;

function TProject.GetConnectionFRC: LongInt;
begin
  Result := 0; // not implemented
end;

function TProject.GetConnectionString: AnsiString;
begin
  Result := 'No connection, not implemented in library!';
end;

end.

